xds/src/main/java/io/grpc/xds/XdsLrsClient.java

```diff
@@ -0,0 +1,325 @@
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Client of XDS load reporting service. It should be run in some serialized context.
+ */
+@NotThreadSafe
+final class XdsLrsClient {
+
+  @VisibleForTesting
+  static final String TRAFFICDIRECTOR_GRPC_HOSTNAME
+      = "com.googleapis.trafficdirector.grpc_hostname";
+  private final String serviceName;
+  private final ManagedChannel channel;
+  private final Helper helper;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timerService;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final Stopwatch stopwatch;
+  private final ChannelLogger logger;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+
+  @Nullable
+  private BackoffPolicy lrsRpcRetryPolicy;
+  @Nullable
+  private ScheduledHandle lrsRpcRetryTimer;
+
+  @Nullable
+  private LrsStream lrsStream;
+
+  XdsLrsClient(ManagedChannel channel, Helper helper, Supplier<Stopwatch> stopwatchSupplier,
+      BackoffPolicy.Provider backoffPolicyProvider) {
+    this.channel = checkNotNull(channel, "channel");
+    this.helper = checkNotNull(helper, "helper");
+    this.serviceName = checkNotNull(helper.getAuthority(), "serviceName");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.stopwatch = stopwatchSupplier.get();
+    this.logger = checkNotNull(helper.getChannelLogger(), "logger");
+    this.timerService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+  }
+
+  // TODO (chengyuanzhang): consider putting to a util class.
+  @VisibleForTesting
+  static long durationToNum(Duration duration, TimeUnit unit) {
+    return unit.convert(duration.getSeconds() * 1000_000_000L + duration.getNanos(),
+        TimeUnit.NANOSECONDS);
+  }
+
+  void startLrsRpc() {
+    checkState(lrsStream == null, "previous lbStream has not been cleared yet");
+    LoadReportingServiceGrpc.LoadReportingServiceStub stub
+        = LoadReportingServiceGrpc.newStub(channel);
+    lrsStream = new LrsStream(stub, stopwatchSupplier.get());
+    lrsStream.start();
+    stopwatch.reset().start();
+    lrsStream.stopwatch.reset().start();
+
+    LoadStatsRequest initRequest =
+        LoadStatsRequest.newBuilder()
+            .setNode(Node.newBuilder()
+                .setMetadata(Struct.newBuilder()
+                    .putFields(
+                        TRAFFICDIRECTOR_GRPC_HOSTNAME,
+                        Value.newBuilder().setStringValue(serviceName).build())))
+            .build();
+    try {
+      lrsStream.lrsRequestWriter.onNext(initRequest);
+    } catch (Exception e) {
```

Comment:
`onNext()` is not expected to throw, unless there is a bug in gRPC. We don't catch it, but instead let it to be thrown to the synchronization context which will put the channel in panic and the user will notice it and file a bug to us.

Reply:
ACK.